File: ok2curl/src/test/java/com/moczul/ok2curl/CookieHandlerTest.java
package com.moczul.ok2curl;

import com.moczul.ok2curl.util.FakeLogger;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okio.BufferedSink;

import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.contains;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class CookieHandlerTest {

    private FakeLogger logger;
    private MockWebServer server;
    private String url;

    @Before
    public void setUpOkHttp() throws IOException {
        logger = mock(FakeLogger.class);

        server = new MockWebServer();
        server.start();
        server.enqueue(new MockResponse().setBody("ok"));
        url = server.url("/test").toString();
    }

    @After
    public void tearDown() throws IOException {
        server.shutdown();
    }

    @Test
    public void testIfRequestBodyThrowsException_returnInfoAboutError() throws Exception {
        final Request request = new Request.Builder().url(url).patch(new RequestBody() {
            @Override
            public MediaType contentType() {
                return MediaType.parse("application/json");
            }

            @Override
            public void writeTo(BufferedSink sink) throws IOException {
                throw new IOException("exception");
            }
        }).build();

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(new CurlInterceptor(logger))
                .cookieJar(getCookieJar())
                .build();

        try {
            okHttpClient.newCall(request).execute();
        } catch (IOException ignore) {}

        final ArgumentCaptor<String> stringArgumentCaptor = ArgumentCaptor.forClass(String.class);
        verify(logger).log(stringArgumentCaptor.capture());
        assertTrue(stringArgumentCaptor.getValue().contains("Error"));
    }

    @Test
    public void testApplicationInterceptor() throws IOException {
        final Request request = new Request.Builder().url(url).build();

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(new CurlInterceptor(logger))
                .cookieJar(getCookieJar())
                .build();

        okHttpClient.newCall(request).execute();

        final String expectedCurl = "curl -X GET " + url;
        verify(logger).log(expectedCurl);
    }

    @Test
    public void testNetworkInterceptor() throws IOException {
        final Request request = new Request.Builder().url(url).build();

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addNetworkInterceptor(new CurlInterceptor(logger))
                .cookieJar(getCookieJar())
                .build();

        okHttpClient.newCall(request).execute();

        verify(logger).log(contains("-H \"Cookie:foo=bar; banana=rama\""));
    }

    private CookieJar getCookieJar() {
        return new CookieJar() {
            @Override
            public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
            }

            @Override
            public List<Cookie> loadForRequest(HttpUrl url) {
                Cookie foo = new Cookie.Builder()
                        .name("foo")
                        .value("bar")
                        .domain(url.host())
                        .build();
                Cookie banana = new Cookie.Builder()
                        .name("banana")
                        .value("rama")
                        .domain(url.host())
                        .build();
                return Arrays.asList(foo, banana);
            }
        };
    }
}

File: ok2curl/src/test/java/com/moczul/ok2curl/CookieHandlerTest.kt
package com.moczul.ok2curl

import com.moczul.ok2curl.logger.Loggable
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okio.BufferedSink
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.contains
import org.mockito.Mockito
import org.mockito.Mockito.verify
import java.io.IOException

class CookieHandlerTest {

    private val logger: Loggable = Mockito.mock(Loggable::class.java)
    private lateinit var server: MockWebServer
    private lateinit var url: String

    @Before
    fun setUpOkHttp() {
        server = MockWebServer()
        server.start()
        server.enqueue(MockResponse().setBody("ok"))
        url = server.url("/test").toString()
    }

    @After
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `should intercept failed request and log error information`() {
        // given
        val request: Request = Request.Builder().url(url).patch(object : RequestBody() {
            override fun contentType(): MediaType? {
                return "application/json".toMediaTypeOrNull()
            }

            override fun writeTo(sink: BufferedSink) {
                throw IOException("exception")
            }
        }).build()
        val okHttpClient = OkHttpClient.Builder()
            .addInterceptor(CurlInterceptor(logger))
            .cookieJar(createCookieJar())
            .build()

        // when
        try {
            okHttpClient.newCall(request).execute()
        } catch (ignore: IOException) {
        }

        // then
        verify(logger).log(contains("Error"))
    }

    @Test
    fun `should intercept request with application interceptor`() {
        // given
        val request: Request = Request.Builder().url(url).build()
        val okHttpClient = OkHttpClient.Builder()
            .addInterceptor(CurlInterceptor(logger))
            .cookieJar(createCookieJar())
            .build()

        // when
        okHttpClient.newCall(request).execute()

        // then
        val expectedCurl = String.format("curl -X GET \"%s\"", url)
        verify(logger).log(expectedCurl)
    }

    @Test
    fun `should intercept request with network interceptor`() {
        // given
        val request: Request = Request.Builder().url(url).build()
        val okHttpClient = OkHttpClient.Builder()
            .addNetworkInterceptor(CurlInterceptor(logger))
            .cookieJar(createCookieJar())
            .build()

        // when
        okHttpClient.newCall(request).execute()

        // then
        verify(logger).log(contains("-H \"Cookie:foo=bar; banana=rama\""))
    }

    private fun createCookieJar(): CookieJar {
        return object : CookieJar {
            override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {}
            override fun loadForRequest(url: HttpUrl): List<Cookie> {
                val foo: Cookie = Cookie.Builder()
                    .name("foo")
                    .value("bar")
                    .domain(url.host)
                    .build()
                val banana: Cookie = Cookie.Builder()
                    .name("banana")
                    .value("rama")
                    .domain(url.host)
                    .build()
                return listOf(foo, banana)
            }
        }
    }
}